tlatools/org.lamport.tlatools/src/tla2sany/semantic/Generator.java

```diff
@@ -2102,11 +2106,145 @@ public final ModuleNode generate(TreeNode treeNode) throws AbortException {
 
 		if (treeNode.isKind(N_Module)) {
 			this.context = symbolTable.getContext();
-			return this.generateModule(treeNode, null);
+			final ModuleNode m = this.generateModule(treeNode, null);
```

Comment:
The generate class is already enormous, and given that this logic only lives here so it can access the top-level module node perhaps it would be better to put it somewhere else - like in a Linter class that is run as a third pass, which this implicitly is.

Reply:
A third pass before or after level checking?

Reply:
I would say after, in case we want to add lints that use level info.

Reply:
Done